pkg/daemon/multus/validation.go

```diff
@@ -0,0 +1,314 @@
+	vsm.SetNextState(&verifyAllClientsReadyState{
+		expectedNumClients: s.expectedNumClients,
+	})
+	return []string{}, nil
+}
+
+// 5. All client pods should be in "Ready" state
+type verifyAllClientsReadyState struct {
+	expectedNumClients int
+
+	// keep some info to heuristically determine if the network might be flaky/overloaded
+	prevNumReady                    int
+	timeClientsStartedBecomingReady time.Time
+	suggestFlaky                    bool
+}
+
+func (s *verifyAllClientsReadyState) Run(ctx context.Context, vsm *validationStateMachine) (suggestions []string, err error) {
+	numReady, err := vsm.vt.numClientsReady(ctx, s.expectedNumClients)
+	collocationSuggestion := "if clients on the same node as the web server become ready but not others, " +
+		"there may be a network firewall or security policy blocking inter-node traffic on multus networks"
+	defaultSuggestions := append([]string{collocationSuggestion, flakyNetworkSuggestion}, unableToProvideAddressSuggestions...)
+	if err != nil {
+		return defaultSuggestions, err
+	}
+
+	s.checkIfFlaky(vsm, numReady)
+
+	if numReady != s.expectedNumClients {
+		return defaultSuggestions, fmt.Errorf("number of ready clients [%d] is not the number expected [%d]", numReady, s.expectedNumClients)
+	}
+
+	vsm.vt.Logger.Infof("all %d clients are ready", s.expectedNumClients)
+	suggestionsOnSuccess := []string{}
+	if s.suggestFlaky {
+		suggestionsOnSuccess = append(suggestionsOnSuccess,
+			fmt.Sprintf("not all clients became ready within %s; %s", flakyThreshold.String(), flakyNetworkSuggestion))
+	}
+	vsm.Exit() // DONE!
+	return suggestionsOnSuccess, nil
+}
+
+// clients should all become ready within a pretty short amount of time since they all should start
+// pretty simultaneously
+var flakyThreshold = 20 * time.Second
+
+func (s *verifyAllClientsReadyState) checkIfFlaky(vsm *validationStateMachine, numReady int) {
+	if s.suggestFlaky {
+		return // no need to do any checks if network is already found flaky
+	}
+
+	if numReady > s.prevNumReady && s.timeClientsStartedBecomingReady.IsZero() {
+		vsm.vt.Logger.Debugf("clients started becoming ready")
+		s.timeClientsStartedBecomingReady = time.Now()
+		return
+	}
+
+	if !s.timeClientsStartedBecomingReady.IsZero() {
+		// check to see how long it took since clients first started becoming ready. if the time is
+		// longer than the flaky threshold, warn the user, and record that the network is flaky
+		if time.Since(s.timeClientsStartedBecomingReady) > flakyThreshold {
+			vsm.vt.Logger.Warningf(
+				"network seems flaky; the time since clients started becoming ready until now is greater than %s", flakyThreshold.String())
+			s.suggestFlaky = true
+		}
+	}
+}
+
+// Run the Multus validation test.
+func (vt *ValidationTest) Run(ctx context.Context) (*ValidationTestResults, error) {
+	if vt.Logger == nil {
+		vt.Logger = &SimpleStderrLogger{}
+		vt.Logger.Infof("no logger was specified; using a simple stderr logger")
```

Comment:
What logger might they specify? Can we just always use the simple logger and not worry about the option?

Reply:
We specify the `capnslog.PackageLogger` for Rook. Chiefly, I include this because I am aware that ocs-operator may want to run this at some point, and it will be easier and a better interaction with the GUI for them if they can call the golang code API instead of executing the binary process.